java-checks/src/main/java/org/sonar/java/checks/HardCodedCredentialsCheck.java

```diff
@@ -43,38 +47,79 @@
 
+  private static final MethodMatcher PASSWORD_AUTHENTICATION_CONSTRUCTOR = MethodMatcher.create()
+    .typeDefinition("java.net.PasswordAuthentication")
+    .name("<init>")
+    .addParameter("java.lang.String")
+    .addParameter("char[]");
+
+  private static final MethodMatcher STRING_TO_CHAR_ARRAY = MethodMatcher.create()
+    .typeDefinition("java.lang.String")
+    .name("toCharArray")
+    .withNoParameterConstraint();
+
   @Override
   public List<Kind> nodesToVisit() {
-    return ImmutableList.of(Tree.Kind.STRING_LITERAL, Tree.Kind.VARIABLE, Tree.Kind.ASSIGNMENT, Tree.Kind.METHOD_INVOCATION);
+    return ImmutableList.of(Tree.Kind.STRING_LITERAL, Tree.Kind.VARIABLE, Tree.Kind.ASSIGNMENT, Tree.Kind.NEW_CLASS, Tree.Kind.METHOD_INVOCATION);
   }
 
   @Override
   public void visitNode(Tree tree) {
     if (tree.is(Tree.Kind.STRING_LITERAL)) {
-      String literalValue = ((LiteralTree) tree).value();
-      if (PASSWORD_LITERAL_PATTERN.matcher(literalValue).find()) {
-        reportIssue(tree);
-      }
+      handleStringLiteral((LiteralTree) tree);
     } else if (tree.is(Tree.Kind.VARIABLE)) {
-      VariableTree variable = (VariableTree) tree;
-      IdentifierTree simpleName = variable.simpleName();
-      if (isStringLiteral(variable.initializer()) && isPasswordVariableName(simpleName)) {
-        reportIssue(simpleName);
-      }
+      handleVariable((VariableTree) tree);
     } else if (tree.is(Tree.Kind.ASSIGNMENT)) {
-      AssignmentExpressionTree assignmentExpression = (AssignmentExpressionTree) tree;
-      ExpressionTree variable = assignmentExpression.variable();
-      if (isStringLiteral(assignmentExpression.expression()) && isPasswordVariable(variable)) {
-        reportIssue(variable);
-      }
+      handleAssignement((AssignmentExpressionTree) tree);
+    } else if (tree.is(Tree.Kind.NEW_CLASS)) {
+      handleConstructor((NewClassTree) tree);
     } else {
-      MethodInvocationTree mit = (MethodInvocationTree) tree;
-      if (isSettingPassword(mit)) {
-        reportIssue((mit).methodSelect());
+      handleMethodInvocation((MethodInvocationTree) tree);
+    }
+  }
+
+  private void handleStringLiteral(LiteralTree tree) {
+    if (PASSWORD_LITERAL_PATTERN.matcher(tree.value()).find()) {
+      reportIssue(tree);
+    }
+  }
+
+  private void handleVariable(VariableTree tree) {
+    IdentifierTree simpleName = tree.simpleName();
+    if (isStringLiteral(tree.initializer()) && isPasswordVariableName(simpleName)) {
+      reportIssue(simpleName);
+    }
+  }
+
+  private void handleAssignement(AssignmentExpressionTree tree) {
+    ExpressionTree variable = tree.variable();
+    if (isStringLiteral(tree.expression()) && isPasswordVariable(variable)) {
+      reportIssue(variable);
+    }
+  }
+
+  private void handleConstructor(NewClassTree tree) {
+    if (PASSWORD_AUTHENTICATION_CONSTRUCTOR.matches(tree)) {
```

Comment:
It's a matter of style, but I would reverse the conditions to not nest 3 `if` statements.